packages/ios-client/Sources/FishjamClient/FishjamClientInternal.swift

```diff
@@ -1,49 +1,568 @@
+                id: eventEndpoint.id, type: EndpointType(fromString: eventEndpoint.type),
+                metadata: eventEndpoint.metadata ?? Metadata())
+            for (trackId, track) in eventEndpoint.tracks {
+                let track = Track(
+                    mediaTrack: nil, endpointId: eventEndpoint.id, rtcEngineId: trackId, metadata: track.metadata)
+                endpoint = endpoint.addOrReplaceTrack(track)
+                listener.onTrackAdded(track: track)
+            }
+            remoteEndpointsMap[eventEndpoint.id] = endpoint
         }
+
+        listener.onJoined(peerID: endpointId, peersInRoom: remoteEndpointsMap)
+        commandsQueue.finishCommand()
+        reconnectionManager?.onReconnected()
     }
 
     func leave() {
-        webrtcClient?.disconnect()
+        rtcEngineCommunication.disconnect()
+        for track in localEndpoint.tracks.values {
+            if let track = track as? LocalTrack {
+                track.stop()
+            }
+        }
+        peerConnectionManager.close()
+        localEndpoint = Endpoint(id: "", type: .WEBRTC)
+        remoteEndpointsMap = [:]
+        peerConnectionManager.removeListener(self)
+        rtcEngineCommunication.removeListener(self)
+        webSocket?.disconnect(closeCode: CloseCode.normal.rawValue)
+        webSocket = nil
         isAuthenticated = false
+        commandsQueue.clear()
     }
 
-    func cleanUp() {
-        webrtcClient?.disconnect()
-        isAuthenticated = false
-        webSocket?.disconnect()
-        webSocket = nil
-        onDisconnected()
+    func createVideoTrack(videoParameters: VideoParameters, metadata: Metadata, captureDeviceName: String? = nil)
+        -> LocalVideoTrack
+    {
+        let videoSource = peerConnectionFactoryWrapper.createVideoSource()
+        let webrtcTrack = peerConnectionFactoryWrapper.createVideoTrack(source: videoSource)
+        let videoCapturer = CameraCapturer(
+            videoParameters: videoParameters, delegate: videoSource, deviceId: captureDeviceName)
+        let videoTrack = LocalVideoTrack(
+            mediaTrack: webrtcTrack, videoSource: videoSource, endpointId: localEndpoint.id,
+            videoParameters: videoParameters,
+            capturer: videoCapturer)
+        videoTrack.start()
+        let promise = commandsQueue.addCommand(
+            Command(commandName: .ADD_TRACK, clientStateAfterCommand: nil) {
+                self.localEndpoint = self.localEndpoint.addOrReplaceTrack(videoTrack)
+                self.peerConnectionManager.addTrack(track: videoTrack)
+                if self.commandsQueue.clientState == .CONNECTED || self.commandsQueue.clientState == .JOINED {
+                    self.rtcEngineCommunication.renegotiateTracks()
+                } else {
+                    self.commandsQueue.finishCommand(commandName: .ADD_TRACK)
+                }
+            })
+        do {
+            try awaitPromise(promise)
+        } catch {
+
```

Comment:
Would it make sense to add something (logs?) in case of such errors?

Reply:
Log added.